File: cpp_raytracing/vec3.cpp
#include <iostream>
#include <math.h>
#include "vec3.h"

// Class vec3 methods

// Constructors
vec3::vec3() {};
vec3::vec3(float e0, float e1, float e2){
    this->e[0] = e0 ; this->e[1] = e1 ; this->e[2] = e2;
};

// Basic access
float vec3::x(){ return this->e[0]; };
float vec3::y(){ return this->e[1]; };
float vec3::z(){ return this->e[2]; };

float vec3::r(){ return this->e[0]; };
float vec3::g(){ return this->e[1]; };
float vec3::b(){ return this->e[2]; };

// Overloading unary operators
vec3 vec3::operator-(){ return vec3( -this->e[0] , -this->e[1], -this->e[2] ); };
float vec3::operator[](int i){ return this->e[i]; };

// Overloading binary operators
vec3 vec3::operator+(vec3 other){
    return vec3( this->e[0] + other[0], this->e[1] + other[1], this->e[2] + other[2] );
};
vec3 vec3::operator-(vec3 other){
    return vec3( this->e[0] - other[0], this->e[1] - other[1], this->e[2] - other[2] );
};
vec3 vec3::operator*(vec3 other){
    return vec3( this->e[0]*other[0], this->e[1]*other[1], this->e[2]*other[2] );
};
vec3 vec3::operator*(float other){
    return vec3( this->e[0]*other, this->e[1]*other, this->e[2]*other );
};
vec3 vec3::operator/(vec3 other){
    return vec3( this->e[0]/other[0], this->e[1]/other[1], this->e[2]/other[2] );
};
vec3 vec3::operator/(float other){
    return vec3( this->e[0]/other, this->e[1]/other, this->e[2]/other );
};

//Other methods
float vec3::length(){
    return sqrt(this->e[0]*this->e[0] +  this->e[1]*this->e[1] + this->e[2]*this->e[2]);
};
float vec3::squared_length(){
    return (this->e[0]*this->e[0] +  this->e[1]*this->e[1] + this->e[2]*this->e[2]);
};

// print
void vec3::print(){
    std::cout << this->e[0] << " " << this->e[1] << " " << this->e[2] << std::endl;
};

//*****************
// Dot and cross products
float dot(vec3 v1, vec3 v2){ return v1[0]*v2[0] + v1[1]*v2[1] + v1[2]*v2[2] ; };
vec3 cross(vec3 v1, vec3 v2){
  float tx = v1.e[1]*v2.e[2] - v1.e[2]*v2.e[1];
  float ty = -(v1.e[0]*v2.e[2] - v1.e[2]*v2.e[0]);
  float tz = v1.e[0]*v2.e[1] - v1.e[1]*v2.e[0];
  return vec3(tx, ty, tz);
};

// Unit vector
vec3 unit_vector(vec3 v){ return v/v.length() ;  };

File: py_raytracing/backup1/vec3.py
#!python3

class vec3:
    def __init__(self, e0, e1, e2):
        self.e = [e0,e1,e2]

    def x(self):
        return self.e[0]
    def y(self):
        return self.e[1]
    def z(self):
        return self.e[2]
    def r(self):
        return self.e[0]
    def g(self):
        return self.e[1]
    def b(self):
        return self.e[2]

    # Overlaod unary operators
    def __neg__(self,other):
        return vec3(-self.e[0], -self.e[1], -self.e[2])
    def __getitem__(self, key):
        return self.e[key]

    # Overload binary operators
    def __add__(self, other):
        return vec3( self.e[0] + other.e[0], self.e[1] + other.e[1], self.e[2] + other.e[2] )
    def __sub__(self, other):
        return vec3( self.e[0] - other.e[0], self.e[1] - other.e[1], self.e[2] - other.e[2] )
    def __mul__(self, other):
        if type(other) != float :
            return vec3( self.e[0]*other.e[0], self.e[1]*other.e[1], self.e[2]*other.e[2] )
        else:
            return vec3( self.e[0]*other, self.e[1]*other, self.e[2]*other )

    def __truediv__(self, other):
        if type(other) != float:
            return vec3( self.e[0]/other.e[0], self.e[1]/other.e[1], self.e[2]/other.e[2] )
        else:
            return vec3( self.e[0]/other, self.e[1]/other, self.e[2]/other )


    # Other methods
    def length(self):
        return (self.e[0]*self.e[0] + self.e[1]*self.e[1] + self.e[2]*self.e[2])**0.5
    def squared_length(self):
        return self.e[0]*self.e[0] + self.e[1]*self.e[1] + self.e[2]*self.e[2]

    def make_unit_vector(self):
        pass


# Dot and cross products
def dot(v1, v2):
    return v1.e[0]*v2.e[0] + v1.e[1]*v2.e[1] + v1.e[2]*v2.e[2]

def cross(v1, v2):
    tx = v1.e[1]*v2.e[2] - v1.e[2]*v2.e[1]
    ty = -(v1.e[0]*v2.e[2] - v1.e[2]*v2.e[0])
    tz = v1.e[0]*v2.e[1] - v1.e[1]*v2.e[0]
    return vec3(tx, ty, tz)

def unit_vector(v):
    return v/v.length()




